addons/stock/models/stock_production_lot.py

```diff
@@ -63,3 +64,8 @@ def action_lot_open_quants(self):
         action = self.env.ref('stock.lot_open_quants').read()[0]
         action['context'] = {'search_default_lot_id': self.id}
         return action
+
+    @api.onchange('product_tmpl_id')
+    def onchange_product_tmpl_id(self):
+        if self.product_tmpl_id:
+            return {'domain':{'product_id': [('product_tmpl_id', '=', self.product_tmpl_id.id)]}}
```

Comment:
Domain will be reset if you edit then save the form by using an onchange to set the domain